### PR TITLE
Implement user checkpoints in CrashMonkey

### PR DESCRIPTION
**Pull request by ashmrtn**
Implementations for #42, #41, #40, #39, and #38.

Checkpoints can now be called from a helper binary in `build/user_tools/Checkpoint` or via the programmatic API with `Checkpoint()`.

The most recent checkpoint reached when generating a crash state is dumped to the output log along with the order of the bios used to generate the crash state (each bio is represented by its index in the original bio stream).

Below is an entry generated by running `sudo ./c_harness -f /dev/vda -d /dev/cow_ram0 -t ext2 tests/checkpoint_example.so`. This test creates a file and a new directory, and then moves the new file between the root directory of the file system and the new directory. Checkpoints are placed at various intervals during this test.

```
 Test #26 FAILED: file missing: test file has completely disappeared
   last checkpoint: 2
   crash state: 0, 1, 2, 2, 3, 4, 4, 5, 6, 7, 7, 8, 9, 17, 16, 18
```

### CODE REVIEW

File: code/tests/checkpoint_example.cpp [hunk 1]
```diff
@@ -162,13 +162,15 @@ class CheckpointExample : public BaseTestCase {
     if (stat_old_res < 0 && errno_old == ENOENT &&
         stat_new_res < 0 && errno_new == ENOENT) {
       test_res->SetError(DataTestResult::kFileMissing);
+      test_res->error_description = "test file has completely disappeared";
```
Comment (vijay03) [inline code/tests/checkpoint_example.cpp]: Can we mention which file has disappeared? So is the file we are testing for always the same?

File: code/tests/checkpoint_example.cpp [hunk 2]
```diff
@@ -221,8 +230,10 @@ class CheckpointExample : public BaseTestCase {
 
     if (bytes_read != strlen(TEST_TEXT)) {
       test_res->SetError(DataTestResult::kFileDataCorrupted);
+      test_res->error_description = check_file + ": read data size wrong";
```
Comment (vijay03) [inline code/tests/checkpoint_example.cpp]: Lets print what the size should be, and what size is reported?

File: code/tests/checkpoint_example.cpp [hunk 3]
```diff
@@ -221,8 +230,10 @@ class CheckpointExample : public BaseTestCase {
...
     } else if (memcmp(TEST_TEXT, buf, strlen(TEST_TEXT)) != 0) {
       test_res->SetError(DataTestResult::kFileDataCorrupted);
+      test_res->error_description = check_file + ": read data incorrect";
```
Comment (vijay03) [inline code/tests/checkpoint_example.cpp]: Similarly, lets print the index/prefix at which the buffers don't match

File: code/tests/checkpoint_example.cpp (hunk 1 above)
Comment (ashmrtn) [reply code/tests/checkpoint_example.cpp]: This error is if the file does not exist at either the old or the new file path (the file is completely gone). I have updated the error string to note the file paths it is checking for the file.

File: code/tests/checkpoint_example.cpp (hunk 2 above)
Comment (ashmrtn) [reply code/tests/checkpoint_example.cpp]: done

File: code/tests/checkpoint_example.cpp (hunk 3 above)
Comment (ashmrtn) [reply code/tests/checkpoint_example.cpp]: done

**Pull request merged**